### PR TITLE
 Redirect from cart to marketplace thank you page

### LINKED ISSUE #74295: Marketplace Themes: Cart redirect to Marketplace Thank you page
Marketplace Themes, after the purchase, should be redirected to the Marketplace Thank you page. See pekYwv-Pz-p2.
The steps to achieve this are: 
 - [ ] Remove the `redirect_to` parameter from the cart URL for Marketplace themes.
 - [ ] Make Marketplace themes enter in the following condition (see https://github.com/Automattic/wp-calypso/issues/73838#issuecomment-1458339666): 

 
https://github.com/Automattic/wp-calypso/blob/3d1006d1780d61012d7a1f1572abd6660cbb6a7d/client/my-sites/checkout/get-thank-you-page-url/index.ts#L526-L535

### PR DESCRIPTION
**Pull request by gcsecsey**
Closes #74295

## Proposed Changes

- Remove the `redirect_to` param from the cart URL for marketplace themes
- update the cart fallback URL logic to get theme and plugin slugs from the `extra` object

#### With an Atomic site
- Put a marketplace theme and a marketplace product into the cart
- Complete checkout
- Check that after checkout, the site redirects to the marketplace thank you page
- Check that in the URL, both the theme and plugin slugs are passed like `?plugins=slug1&themes=slug2`

CleanShot 2023-03-21 at 15 52 29@2x

#### With a non-atomic site
- Put a marketplace theme in the cart
- In checkout, check that both the theme and a plan is added to the cart
- Complete checkout
- Check that the site redirects back to the theme page and the CTA is fetching the Atomic transfer status as implemented in https://github.com/Automattic/wp-calypso/pull/72837

CleanShot 2023-03-21 at 16 00 03@2x

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: packages/shopping-cart/src/types.ts [hunk 1]
```diff
@@ -564,7 +564,8 @@ export interface ResponseCartProductExtra {
 
 	// Marketplace properties
 	is_marketplace_product?: boolean;
-	plugin_slug?: boolean;
+	product_slug?: boolean;
+	product_type?: string;
```
Comment (sirbrillig) [inline packages/shopping-cart/src/types.ts]: Could we add a comment above this to explain what `product_type` and `product_slug` are for and when they're likely to be set? I know there's a lot of uncommented properties here but I'm trying to add comments for new ones to help us in the future.

File: packages/shopping-cart/src/types.ts [hunk 2]
```diff
@@ -564,7 +564,8 @@ export interface ResponseCartProductExtra {
 
 	// Marketplace properties
 	is_marketplace_product?: boolean;
-	plugin_slug?: boolean;
+	product_slug?: boolean;
```
Comment (WBerredo) [inline packages/shopping-cart/src/types.ts]:
```suggestion
	product_slug?: string;
```

This should be string, right? I think the current version is wrong.

File: client/my-sites/checkout/get-thank-you-page-url/index.ts [hunk 3]
```diff
@@ -533,14 +533,23 @@ function getFallbackDestination( {
...
+		cart?.products?.filter( ( product ) => product?.extra?.is_marketplace_product ) || [];
 
-	if ( pluginSlugs.length > 0 ) {
+	const marketplacePluginSlugs = marketplaceProducts
+		?.filter( ( { extra } ) => extra.product_type === 'marketplace_plugin' )
+		.map( ( { extra } ) => extra.product_slug );
+
+	const marketplaceThemeSlugs = marketplaceProducts
+		?.filter( ( { extra } ) => extra.product_type === 'marketplace_theme' )
+		.map( ( { extra } ) => extra.product_slug );
```
Comment (WBerredo) [inline client/my-sites/checkout/get-thank-you-page-url/index.ts]:
```suggestion
	const marketplacePluginSlugs = marketplaceProducts
		.filter( ( { extra } ) => extra.product_type === 'marketplace_plugin' )
		.map( ( { extra } ) => extra.product_slug );

	const marketplaceThemeSlugs = marketplaceProducts
		.filter( ( { extra } ) => extra.product_type === 'marketplace_theme' )
		.map( ( { extra } ) => extra.product_slug );
```

Nit: As `marketplaceProducts` will always be an array (maybe empty but still) I think you can remove the conditional check (`?`)

**Review by WBerredo — COMMENTED** [review]
For the non-atomic case, I think we must turn the site into atomic on the Thank You page as we do for plugins. (I just saw your comment on the ticket later, we said the same thing 😀)

**Comment by WBerredo** [comment]
I added the `DO NOT MERGE` tag as this can only be merged when all the work on the Thank You page is done to support Themes.

File: packages/shopping-cart/src/types.ts (hunk 2 above)
Comment (gcsecsey) [reply packages/shopping-cart/src/types.ts]: Ooops, yes it should, I didn't notice that one!

File: client/my-sites/checkout/get-thank-you-page-url/index.ts (hunk 3 above)
Comment (gcsecsey) [reply client/my-sites/checkout/get-thank-you-page-url/index.ts]: Thanks for pointing this out, I was planning to remove this at one point but received a TS error, went back to fix it, then forgot to remove the optional chaining.

File: packages/shopping-cart/src/types.ts (hunk 1 above)
Comment (gcsecsey) [reply packages/shopping-cart/src/types.ts]: Thanks @sirbrillig, it's a good idea to make new properties easier to understand. I've updated the types to make them more representative and added comments to them.

**Comment by gcsecsey** [comment]
> For the non-atomic case, I think we must turn the site into atomic on the Thank You page as we do for plugins. (I just saw your comment on the ticket later, we said the same thing 😀)

Thanks for the confirmation @WBerredo. In that case, I'll remove the `redirect` parameter from the checkout entirely and update the test steps.

**Comment by gcsecsey** [comment]
I've removed the parameter, now all sites are redirected to the new thank you page.

**Review by epeicher — APPROVED** [review]
I have tested it, and it works as described in the comments. Both Atomic and non-Atomic sites are redirected to the Thank You page. LGTM!

**Pull request merged**